File: LeetCode/Python/Cousins in Binary Tree II.py
"""
  Cousins in Binary Tree II
  URL: https://leetcode.com/problems/cousins-in-binary-tree-ii
  Difficulty: Medium
  Topic : Hash Table, Tree, Depth-First Search, Breadth-First Search, Binary Tree
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def replaceValueInTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        depthSum = root.val
        nextSum = [root.val]
        cur = [[root]]
        next = []
        while len(cur) > 0:
            newDepthSum = 0
            newNextSum = []
            for i in range(len(cur)):
                for node in cur[i]:
                    node.val = depthSum - nextSum[i]
                    pair = []
                    pairSum = 0
                    if node.left != None:
                        pair.append(node.left)
                        pairSum += node.left.val
                    if node.right != None:
                        pair.append(node.right)
                        pairSum += node.right.val
                    next.append(pair)
                    newNextSum.append(pairSum)
                    newDepthSum += pairSum
            cur = next
            next = []
            nextSum = newNextSum
            depthSum = newDepthSum
        return root

File: LeetCode/Medium/Cousins in Binary Tree II.swift
/*
  Cousins in Binary Tree II
  URL: https://leetcode.com/problems/cousins-in-binary-tree-ii
  Difficulty: Medium
  Topic : Hash Table, Tree, Depth-First Search, Breadth-First Search, Binary Tree
*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func replaceValueInTree(_ root: TreeNode?) -> TreeNode? {
        var depthSum = root!.val
        var nextSum = [root!.val]
        var cur: [[TreeNode]] = [[root!]]
        var next: [[TreeNode]] = []
        while !cur.isEmpty {
            var newDepthSum = 0
            var newNextSum: [Int] = []
            for i in 0..<cur.count {
                for node in cur[i] {
                    node.val = depthSum - nextSum[i]
                    var pair: [TreeNode] = []
                    var sum = 0
                    if let left = node.left {
                        pair.append(left)
                        sum += left.val
                    }
                    if let right = node.right { 
                        pair.append(right)
                        sum += right.val
                    }
                    next.append(pair)
                    newNextSum.append(sum)
                    newDepthSum += sum
                }
            }
            cur = next
            next = []
            nextSum = newNextSum
            depthSum = newDepthSum
        }
        return root
    }
}
